cvmfs/catalog_mgr_rw_impl.h

```diff
@@ -0,0 +1,181 @@
+  virtual_node_t root_node(catalog_path, catalog_mgr_);
+  root_node.ExtractChildren(catalog_mgr_);
+  // we have just recursively loaded the entire virtual tree!
+  OptimalPartition(&root_node);
+}
+
+template <class CatalogMgrT>
+void CatalogBalancer<CatalogMgrT>::OptimalPartition(
+    virtual_node_t *virtual_node) {
+  // postorder track of the fs-tree
+  for (unsigned i = 0; i < virtual_node->children.size(); ++i) {
+    virtual_node_t *virtual_child =
+        &virtual_node->children[i];
+    if (virtual_child->IsDirectory() && !virtual_child->IsCatalog())
+      OptimalPartition(virtual_child);
+  }
+  virtual_node->CaltulateWeight();
+  while (virtual_node->weight > catalog_mgr_->balance_weight_) {
+    typename CatalogBalancer<CatalogMgrT>::VirtualNode *heaviest_node =
+        MaxChild(virtual_node);
+    // we directly add a catalog in this node
+    // TODO(molina) apply heuristics here
+    if (heaviest_node != NULL &&
+        heaviest_node->weight >= catalog_mgr_->min_weight_) {
+      AddCvmfsCatalogFile(heaviest_node->path);
+      AddCatalog(heaviest_node);
+      virtual_node->CaltulateWeight();
+    } else {
+      // there is no possibility for this directory to be a catalog
+      break;
+    }
+  }
+}
+
+template <class CatalogMgrT>
+typename CatalogBalancer<CatalogMgrT>::VirtualNode*
+CatalogBalancer<CatalogMgrT>::MaxChild(
+    virtual_node_t *virtual_node)
+{
+  virtual_node_t *max_child = NULL;
+  unsigned max_weight = 0;
+  if (virtual_node->IsDirectory() &&
+      !virtual_node->IsCatalog() &&
+      !virtual_node->is_new_nested_catalog) {
+    for (unsigned i = 0; i < virtual_node->children.size(); ++i) {
+      virtual_node_t *child = &virtual_node->children[i];
+      if (child->IsDirectory() &&
+          !child->IsCatalog() &&
+          max_weight < child->weight) {
+        max_weight = child->weight;
+        max_child = child;
+      }
+    }
+  }
+  return max_child;
+}
+
+template <class CatalogMgrT>
+void CatalogBalancer<CatalogMgrT>::AddCatalog(virtual_node_t *child_node) {
+  if (child_node != NULL) {
+    string new_catalog_path = child_node->path.substr(1);
+    catalog_mgr_->CreateNestedCatalog(new_catalog_path);
+    child_node->weight = 1;
+    child_node->is_new_nested_catalog = true;
+    LogCvmfs(kLogPublish, kLogStdout, "automatic creation of nested"
+        " catalog in '%s'", child_node->path.c_str());
+  }
+}
+
+template <class CatalogMgrT>
+void CatalogBalancer<CatalogMgrT>::VirtualNode::ExtractChildren(
+    CatalogMgrT *catalog_mgr) {
+  DirectoryEntryList direntlist;
+  catalog_mgr->Listing(path, &direntlist);
+  for (unsigned i = 0; i < direntlist.size(); ++i) {
+    string children_path = path + "/" + direntlist[i].name().ToString();
+    children.push_back(virtual_node_t(
+        children_path, direntlist[i], catalog_mgr));
+    weight += children[i].weight;
+  }
+}
+
+template <class CatalogMgrT>
+void CatalogBalancer<CatalogMgrT>::VirtualNode::CaltulateWeight() {
```

Comment:
Hmmmm it is not really propagating weights, it is fixing them. Actually in the python code this function is called fix_weight().
This function is called in the father when one of its children has changed its weight. This phenomenon only happens when one of his children has become a new autogenerated nested catalog, and its weight is now 1 (which represents the sole DirectoryEntry of that directory).
However this is not propagated to the top or the bottom of the tree, but each VirtualNode (only directories, of course) is responsible for calling it when previous operations might have changed the weight of its children (and consequently its own weight).